### PR TITLE
Dynamic engine suspend/resume via prefill.

### PR DESCRIPTION
**Pull request by lmcafee-nvidia**
# What does this PR do ?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: tests/unit_tests/inference/engines/test_dynamic_engine.py [hunk 1]
```diff
@@ -119,6 +120,9 @@ class DynamicEngineTestEnv:
     config: DynamicEngineTestConfig
     requests: List[DynamicInferenceRequest]
     engine: DynamicInferenceEngine
+    mem_usage: dict = field(
+        default_factory=lambda: {"start": None, "end": None, "suspend_resume": {}}
+    )
```
Comment (tdene) [inline tests/unit_tests/inference/engines/test_dynamic_engine.py:139]: Needs to be finished; right now only `end` is tracked.

File: megatron/core/inference/engines/dynamic_engine.py [hunk 2]
```diff
@@ -86,6 +89,12 @@ class DynamicInferenceEngine(AbstractEngine):
         termination_id (int): Token ID to mark end-of-sequence.
         random_seed (Optional[int]): Use a random seed if you want deterministic
             results. Defaults to None.
+        unified_memory_level (Optional[int]): Set unified memory usage within the
+            dynamic inference context. The levels are: 0) no unified memory, 1)
+            allocate `memory_buffer` only in unified memory, and 2) allocate all
+            tensors in unified memory. Note that levels 1 and 2 not not perform
+            any deallocations and allocations when suspending and resuming the
+            context.
```
Comment (tdene) [inline megatron/core/inference/engines/dynamic_engine.py]: I don't like this. `unified_memory_level` is already an argument in the dynamic context - which is where it should go. The engine should just take the context's argument.

File: megatron/core/inference/engines/dynamic_engine.py [hunk 3]
```diff
@@ -129,16 +161,22 @@ def __init__(
...
         self._loop = get_asyncio_loop()
         self._cond = asyncio.Condition()
 
+    def create_cuda_graphs(self):
+        """Create cuda graphs.
+
+        This method iterates the dynamic context's `cuda_graph_request_counts`
+        to record and capture cuda graphs.
+        """
+
```
Comment (tdene) [inline megatron/core/inference/engines/dynamic_engine.py]: Personally, I need this `create_cuda_graphs` method for 2 other PR chains.

I would like to factor it out in an entirely separate PR, that's just code refactoring no changes, and hopefully gets merged more quickly. I'll ping you if I manage to do that.

File: megatron/core/inference/engines/dynamic_engine.py [hunk 4]
```diff
@@ -358,6 +396,184 @@ async def start_listening_to_data_parallel_coordinator(
...
+            delete_cuda_graphs()
+
+    def resume(self):
+        """Resume engine by reallocating context's GPU state."""
+
+        # Skip if not suspended, which can happen when using the inference
+        # coordinator.
+        if not self.is_suspended:
+            return
+        self.is_suspended = False
```
Comment (tdene) [inline megatron/core/inference/engines/dynamic_engine.py:567]: Put a check in `async_step` to prevent stepping while the engine is suspended. And make the condition on L1164 inside `run_engine` also check for `not self.is_suspended`, then run `_notify_cond_for_new_request` (examples elsewhere in this file) upon request to wake up the `run_engine` condition.

File: megatron/core/inference/engines/dynamic_engine.py [hunk 5]
```diff
@@ -358,6 +396,184 @@ async def start_listening_to_data_parallel_coordinator(
...
+                    (
+                        request.prompt_tokens,
+                        torch.tensor(
+                            request.generated_tokens,
+                            dtype=request.prompt_tokens.dtype,
+                            device=request.prompt_tokens.device,
+                        ),
+                    ),
+                    dim=0,
+                )
```
Comment (tdene) [inline megatron/core/inference/engines/dynamic_engine.py]: I think it is very bad if suspending/resuming the engine misaligns the boundary between prompt and generated tokens. That boundary should be saved and restored as well.
For RL, as an example use-case, that boundary needs to be preserved. The right layer of abstraction to do so is here in the engine.

The best way to do so to me, is to keep the old boundary around in the metadata. Make a new `inference_request.Status` with the key of `SUSPENDED`. And use `post_process_requests` to handle the transition between `SUSPENDED` and active, which includes restoring the old boundary after we're doing recomputing the prefill.

File: megatron/core/inference/engines/dynamic_engine.py [hunk 6]
```diff
@@ -975,9 +1193,13 @@ async def run_engine_with_coordinator(
                 # todo [Siddharth]: Can this hardcoded sleep be avoided
                 # with asyncio zmq sockets?
                 if self.paused:
+                    self.suspend()
```
Comment (tdene) [inline megatron/core/inference/engines/dynamic_engine.py:1310]: @sidsingh-nvidia can you comment here?

My understanding is that "paused" and "suspended" are different states. "paused" means that request data continues to sit in the context's paused memory block. "suspended" means that request data is out of the context's memory altogether.

I'm not sure if I'm correct about my interpretation though.

File: megatron/core/inference/engines/dynamic_engine.py [hunk 7]
```diff
@@ -129,16 +161,22 @@ def __init__(
         self.step_end_event = torch.cuda.Event(enable_timing=True)
         self.paused = False
         self.stopped = False
-        self.enable_chunked_prefill = enable_chunked_prefill
 
         # Initialize the asyncio loop if it has not already been initialized.
         # TODO: Start the engine loop here.
         self._loop = get_asyncio_loop()
```
Comment (tdene) [inline megatron/core/inference/engines/dynamic_engine.py]: I just remembered: Robert and I had issues with loops in a similar circumstance. You have the engine running in a loop, you need to reset it, and you get a new loop in the process.

`self._loop = get_asyncio_loop()` might not return the same loop on every pass, which will result in hard-to-debug lockups. For Robert and I, we also had other asyncio things happening at the same time that might have caused the wrong event loop to be returned.

But just in case, I've opened #1990. Overall, that PR should make the event loops safer to use. If you'd like to use that PR, do a `self._loop = None` in `__init__` followed by a `self._loop = get_asyncio_loop(self._loop)` in `reset`.

Comment (tdene) [reply megatron/core/inference/engines/dynamic_engine.py]: Feel free to not use #1990 but I'm hoping that PR will get merged within the next 12 hours.

File: megatron/core/inference/engines/dynamic_engine.py (hunk 3 above)
Comment (tdene) [reply megatron/core/inference/engines/dynamic_engine.py]: I made the PR in question (separate `create_cuda_graphs` method) at #1992. You're listed as co-author on the CG commit; let me know if that's okay.    
It may take a while to merge. I don't think you should wait for it.

I will finish writing up an RFC that involves #1992 and discuss it with the inference team. If this #1982 gets merged first, I'll fix the conflict with #1992. If #1992 gets merged first, I can fix the conflict with this #1982 if you give me write permission.

File: megatron/core/inference/engines/dynamic_engine.py (hunk 6 above)
Comment (tdene) [reply megatron/core/inference/engines/dynamic_engine.py:1310]: Indeed, I now remember better the difference between "paused" and "suspended".

"paused" just means "this engine will no longer step". That's the one and only guarantee it provides.
It was implemented to support RL-like workloads. It's what RL uses for the non-recompute case.

"suspended" seems like it means "this engine no longer has valid KV$".

This is a deeper discussion to be had with @ArEsKay3. I will try to set up for that discussion. But until then, I'm fairly certain that "paused" and "suspended" must exist as separate states. A rank may be "paused" but not "suspended", and honestly, I think it is fair to say that a rank may also be "suspended" but not "paused".

File: megatron/core/inference/engines/dynamic_engine.py (hunk 7 above)
Comment (tdene) [reply megatron/core/inference/engines/dynamic_engine.py]: Now that #1990 is merged into main, I've made a PR into your branch. See [#4](https://github.com/lmcafee-nvidia/Megatron-LM/pull/4).

**Comment by chtruong814** [comment]
@lmcafee-nvidia there's some merge conflicts that need to be resolved

File: tests/unit_tests/inference/engines/test_dynamic_engine.py (hunk 1 above)
Comment (lmcafee-nvidia) [reply tests/unit_tests/inference/engines/test_dynamic_engine.py:139]: the result of a bad rebase a long time ago, fixed now!

File: megatron/core/inference/engines/dynamic_engine.py (hunk 3 above)
Comment (tdene) [reply megatron/core/inference/engines/dynamic_engine.py]: @lmcafee-nvidia looks like #1992 got merged and reimplemented this logic.

I expect to have a PR into your branch in 3 hours from now, that merges this branch and main.

Comment (tdene) [reply megatron/core/inference/engines/dynamic_engine.py]: Done in [!5](https://github.com/lmcafee-nvidia/Megatron-LM/pull/5).

**Comment by kvareddy** [comment]
@sidsingh-nvidia @santhnm2 @shanmugamr1992 , can you please review this MR

File: megatron/core/inference/engines/dynamic_engine.py [hunk 8]
```diff
@@ -362,6 +392,166 @@ async def start_listening_to_data_parallel_coordinator(
...
+            add_time = time.time() - add_time
+
+        # Print inner timing (must be outside context manager above for correct formatting).
+        logging.info(", ".join((
+            f"    > inner timing: alloc {alloc_time:.3f}",
+            f"add {add_time:.3f}",
+            f"capture {capture_time:.3f}.",
+        )))
+
+        return futures
```
Comment (tdene) [inline megatron/core/inference/engines/dynamic_engine.py]: This requires some care.

The script that running the engine has already obtained futures for its submitted requests. We cannot let those futures be invalidated. We have to make sure the old future objects are the ones that get hit with `set_result` once the request finishes.

In the current code, running `add_request` overwrites the old future object with a new one. We need to prevent that, and have it just leave the old one in place.

File: megatron/core/inference/engines/dynamic_engine.py (hunk 5 above)
Comment (sidsingh-nvidia) [reply megatron/core/inference/engines/dynamic_engine.py]: Can't we store this boundary at the client level?

File: megatron/core/inference/engines/dynamic_engine.py (hunk 8 above)
Comment (sidsingh-nvidia) [reply megatron/core/inference/engines/dynamic_engine.py]: Are we using these futures? or the ones created by InferenceClient?

**Review by sidsingh-nvidia — APPROVED** [review]
This has addressed my comment wrt dealing with coordinator signals

File: megatron/core/inference/engines/dynamic_engine.py (hunk 8 above)
Comment (sidsingh-nvidia) [reply megatron/core/inference/engines/dynamic_engine.py]: atleast in my inference-coordinator/client workflow the inference-client created futures are awaited upon, rather than the engine futures
.

Comment (lmcafee-nvidia) [reply megatron/core/inference/engines/dynamic_engine.py]: I'm now reusing the original futures. @tdene please take another look and let me know if it looks okay

File: megatron/core/inference/engines/dynamic_engine.py (hunk 6 above)
Comment (lmcafee-nvidia) [reply megatron/core/inference/engines/dynamic_engine.py:1310]: @tdene , I separated the pause/unpause and suspend/resume signals. The other relevant update is in the client here: https://github.com/NVIDIA/Megatron-LM/pull/1982/files#diff-a0e6e75f00de21c071f50f7c36cf920b4e25b194ed752bd6fb4f122c4cd3ea39R177

File: megatron/core/inference/engines/dynamic_engine.py (hunk 2 above)
Comment (lmcafee-nvidia) [reply megatron/core/inference/engines/dynamic_engine.py]: You are right. I removed the arg and am now grabbing this from the context.

File: megatron/core/inference/engines/dynamic_engine.py (hunk 4 above)
Comment (lmcafee-nvidia) [reply megatron/core/inference/engines/dynamic_engine.py:567]: @tdene , I added these checks. Please double check my usage of `self._cond` and `self._notify_cond_for_new_request()` at the end of `def resume()` here: https://github.com/NVIDIA/Megatron-LM/pull/1982/files#diff-fc07a9e1656a5b83c86e3b0224a343a460ee38c0b0f176c579deec5a284f66f4R550

File: megatron/core/inference/engines/dynamic_engine.py (hunk 5 above)
Comment (lmcafee-nvidia) [reply megatron/core/inference/engines/dynamic_engine.py]: @tdene @sidsingh-nvidia , I'm now storing what I call `DynamicInferenceRequestRecord` in the engine [here](https://github.com/NVIDIA/Megatron-LM/pull/1982/files#diff-fc07a9e1656a5b83c86e3b0224a343a460ee38c0b0f176c579deec5a284f66f4R550), where this struct contains the history of a given request across multiple suspends, as defined [here](https://github.com/NVIDIA/Megatron-LM/pull/1982/files#diff-c9aa579668ef4ae7e658fd0c8964e290755e665e529bbcdd6009256cc577ea33R317).

Calling `record.suspend()` [here](https://github.com/NVIDIA/Megatron-LM/pull/1982/files#diff-c9aa579668ef4ae7e658fd0c8964e290755e665e529bbcdd6009256cc577ea33R335) makes the record store the previous version of the request and then append a new version that concatenates the previous prompt and generated tokens.

Please let me know if this looks good.

File: megatron/core/inference/inference_request.py [hunk 9]
```diff
@@ -314,6 +314,105 @@ def failed(self) -> bool:
...
+        prompt_tokens = self.requests[0].prompt_tokens
+        generated_tokens = merge_lists("generated_tokens")
+
+        # Merged request.
+        request = DynamicInferenceRequest(
+            request_id=self.requests[0].request_id,
+            prompt=tokenizer.detokenize(prompt_tokens.tolist()),
+            prompt_tokens=prompt_tokens,
+            prompt_log_probs=merge_lists("prompt_log_probs"),
+            prompt_top_n_logprobs=merge_lists("prompt_top_n_logprobs"),
```
Comment (tdene) [inline megatron/core/inference/inference_request.py]: I think these two will be wrong. Each suspend/resume pass will overwrite this, and this will end up containing decode-phase log_probs as well.

The only good way I can think of to fix this is along the lines of `prompt_len = len(prompt_tokens)`, `prompt_log_probs=merge_lists("prompt_log_probs")[:prompt_len]`, `prompt_top_n_log_probs=merge_lists("prompt_top_n_log_probs")[:prompt_len]`.

Comment (lmcafee-nvidia) [reply megatron/core/inference/inference_request.py]: this was my mistake, I think these should simply be:

- `prompt_log_probs=self.requests[0].prompt_log_probs`
- `prompt_top_n_logprobs=self.requests[0].prompt_top_n_logprobs`

right?

File: tests/unit_tests/inference/engines/test_dynamic_engine.py [hunk 10]
```diff
@@ -849,24 +895,77 @@ def test_events(self):
...
 
+    @pytest.mark.internal
+    @pytest.mark.skipif(
+        not is_fa_min_version("2.7.3"), reason="need latest flash attn for dynamic batching"
+    )
+    def test_suspend_resume_memory(self):
+
+        # Run tests.
+        mem_usages = {}
+        for suspend_resume_interval in None, 4, 2, 1:
```
Comment (tdene) [inline tests/unit_tests/inference/engines/test_dynamic_engine.py]: Nit: make this a test parameter with `pytest.mark.parametrize`?

File: megatron/core/inference/inference_request.py (hunk 9 above)
Comment (lmcafee-nvidia) [reply megatron/core/inference/inference_request.py]: just pushed this update, let me know how it looks

Comment (tdene) [reply megatron/core/inference/inference_request.py]: Oh yeah, you're absolutely right. Slipped my mind.

**Review by shanmugamr1992 — APPROVED** [review]
